data/parameters.md
withParameters(gl, {
  blend: false,
  blendColor: [0, 0, 0, 0],
  blendEquation: [GL.FUNC_ADD, GL.FUNC_ADD], // [GL.BLEND_EQUATION_RGB, GL.BLEND_EQUATION_ALPHA],
  blendFunc: [GL.ONE, GL.ZERO, GL.ONE, GL.ZERO],

  clearColor: [0, 0, 0, 0],
  colorMask: [true, true, true, true],
  colorWritemask: ,

  cullFace: false,
  cullFaceMode: GL.BACK,

  depthTest: false,
  depthClearValue: 1,
  depthFunc: GL.LESS,
  depthRange: [0, 1],
  depthWritemask: true,

  dither: true,

  frontFace: GL.CCW,

  generateMipmapHint: GL.DONT_CARE,

  lineWidth: 1,

  polygonOffsetFill: false,
  polygonOffset: [0, 0],

  sampleCoverage: GL_SAMPLE_COVERAGE,

  scissorTest: false,
  scissorBox: [0, 0, 1024, 1024],

  stencilTest: false,
  stencilClearValue: 0,
  stencilMask: [0xFFFFFFFF, 0xFFFFFFFF],
  stencilFunc: [GL.ALWAYS, 0, 0xFFFFFFFF, GL.ALWAYS, 0, 0xFFFFFFFF],
  stencilOp: [GL.KEEP, GL.KEEP, GL.KEEP, GL.KEEP, GL.KEEP, GL.KEEP],

  viewport: [0, 0, 1024, 1024],

  [GL.PACK_ALIGNMENT: 4,
  [GL.UNPACK_ALIGNMENT: 4,
  [GL.UNPACK_FLIP_Y_WEBGL]: false,
  [GL.UNPACK_PREMULTIPLY_ALPHA_WEBGL]: GL.UNPACK_PREMULTIPLY_ALPHA_WEBGL,
  [GL.UNPACK_COLORSPACE_CONVERSION_WEBGL]: GL.BROWSER_DEFAULT_WEBGL,

  // WEBGL2 PIXEL PACK/UNPACK MODES

  [GL.PACK_ROW_LENGTH: 0,
  [GL.PACK_SKIP_PIXELS: 0,
  [GL.PACK_SKIP_ROWS: 0,
  [GL.UNPACK_ROW_LENGTH: 0,
  [GL.UNPACK_IMAGE_HEIGHT: 0,
  [GL.UNPACK_SKIP_PIXELS: 0,
  [GL.UNPACK_SKIP_ROWS: 0,
  [GL.UNPACK_SKIP_IMAGES: 0

---

| `blend` | `false` | |
| `blendColor` | `new Float32Array([0, 0, 0, 0])` | |
| `blendEquation` | `[GL.FUNC_ADD, GL.FUNC_ADD]` | |
| `blendFunc` | `[GL.ONE, GL.ZERO, GL.ONE, GL.ZERO]` | |

| `clearColor` | `new Float32Array([0, 0, 0, 0])` | |
| `colorMask` | `[true, true, true, true]` | |

| `cullFace` | `false` | |
| `cullFaceMode` | `GL.BACK` | |

| `depthTest` | `false` | |
| `depthClearValue` | `1` | |
| `depthFunc` | `GL.LESS` | |
| `depthRange` | `new Float32Array([0, 1])` | |
| `depthWritemask` | `true` | |

| `dither` | `true` | |

| `fragmentShaderDerivativeHint` | `GL.DONT_CARE` | |

| `frontFace` | `GL.CCW` | |
| `generateMipmapHint` | `GL.DONT_CARE, // Hint for quality of images generated with` glGenerateMipmap |

| `lineWidth` | `1` | No longer supported by most browsers |

| `polygonOffsetFill` | `false` | Add small offset to fragment depth values |
| `polygonOffset` | `[0, 0]` | factor × DZ + r × units |

| `sampleCoverage` | 
| GL_SAMPLE_ALPHA_TO_COVERAGE |
| GL_SAMPLE_COVERAGE | specify multisample coverage parameters |

| `sampleCoverage` | `[1.0, false]`,
| `scissorTest` | `false` | |
| `scissorBox` | `new Int32Array([0, 0, 1024, 1024])` | |
| `stencilTest` | `false` | |
| `stencilClearValue` | `0` | Sets index used when stencil buffer is cleared |
| `stencilMask` | `[0xFFFFFFFF, 0xFFFFFFFF]` | Sets bit mask enabling writing of individual bits in the stencil planes |
| `stencilFunc` | `[GL.ALWAYS, 0, 0xFFFFFFFF, GL.ALWAYS, 0, 0xFFFFFFFF]` | Set stencil testing function, reference value and mask for front and back |
| `stencilOp` | `[GL.KEEP, GL.KEEP, GL.KEEP, GL.KEEP, GL.KEEP, GL.KEEP]` | Specifies the action to take when the stencil test fails, front and back |
| viewport | new Int32Array([0, 0, 1024, 1024]) | |

| WEBGL1 PIXEL PACK/UNPACK MODES | | |

| [GL.PACK_ALIGNMENT] | 4,   // Packing of pixel data in memory (1,2,4,8)
| [GL.UNPACK_ALIGNMENT] | 4,   // Unpacking pixel data from memory(1,2,4,8)
| [GL.UNPACK_FLIP_Y_WEBGL]: false, // Flip source data along its vertical axis
| [GL.UNPACK_PREMULTIPLY_ALPHA_WEBGL]:GL.UNPACK_PREMULTIPLY_ALPHA_WEBGL, // Multiplies the alpha channel into the other color channels
| [GL.UNPACK_COLORSPACE_CONVERSION_WEBGL]:GL.BROWSER_DEFAULT_WEBGL, // Default color space conversion or no color space conversion.

WEBGL2 PIXEL PACK/UNPACK MODES


| [GL.PACK_ROW_LENGTH] | 0,   // Number of pixels in a row.
| [GL.PACK_SKIP_PIXELS] | 0,// Number of pixels skipped before the first pixel is written into memory.
| [GL.PACK_SKIP_ROWS] | 0, // Number of rows of pixels skipped before first pixel is written to memory.
| [GL.UNPACK_ROW_LENGTH] | 0, // Number of pixels in a row.
| [GL.UNPACK_IMAGE_HEIGHT] | 0, // Image height used for reading pixel data from memory
| [GL.UNPACK_SKIP_PIXELS] | 0, // Number of pixel images skipped before first pixel is read from memory
| [GL.UNPACK_SKIP_ROWS] | 0, // Number of rows of pixels skipped before first pixel is read from memory
| [GL.UNPACK_SKIP_IMAGES] | 0 | Number of pixel images skipped before first pixel is read from memory


CHANGELOG.md
# Pre-releases

## 4.0.0-alpha.

- Un-deprecate `scenegraph` module (except `Scene`), merge with `core` module.
- `shadertools` module no longer experimental

## 4.0.0-alpha.10

- FIX: Framebuffer resize & add unit test (#200)
- Add the pixel parameter back in texture class for compatibility with v3 (#198)
- FIX: 'npm run build' for demos (#195)
- FIX: Lesson 08 (#196)
- FIX: Lesson 07 (#194)

## 4.0.0-alpha.9

- FIX: Some leftover export fix and storage mode fix (#192)

## 4.0.0-alpha.8

- FIX: Remove duplicate export that fails deck.gl tests (#191)

## 4.0.0-alpha.7

- FIX: Fix the texture storage mode settings (#189)
- FIX: examples/lessons (#188)
- Transform feedback fixes (#187)
- FIX: Example updates and fixes for textures (#186)
- Size improvements to transpiled code (dist)
- Tree-shaking improvements - carefully avoid dependencies that defeat tree shaking (#185)

## 4.0.0-alpha.6
- Fix framebuffer creation error AGAIN (#183)
- NEW: `UniformBufferLayout` class

## 4.0.0-alpha.4
- Remove null params given to Float32Array constructor (#176)
- Fix framebuffer creation error (#177)

## 4.0.0-alpha.3
- bump seer
- remove duplicate info from readme

## 4.0.0-alpha.2

- Reorgnize files (#168)
- Transform feedback improvement (#165)
- WebGL2 updates (#160)
- Buffer refactor (#156)
- Fix examples (#161, #149, #172, #173)
- Adding new docs for WebGL2 (#159)
- Demo site creation (#158)
- Docs cleanup and updates (#157, #169, #170)
- seer integration
- Add coverage support (#155)

## 4.0.0-alpha.1

- Refactor WebGL classes using new `Resource` base class
- `Resource.getParameters` for ease of debugging
- Fix FramebufferObject export
- GL state and limit management (#146)
- Fix shader file name (#151)
- Refactor many classes in the webgl folder (#136, #154)
- Check compilation and linking status only with debug WebGL context (#144)
- Add benchmarking scaffolding and a benchmark test for Program constructor (#142)
- Docs update (#137)


# Official releases


## v3

Theme: Pure ES6 Codebase/Build tooling improvements

### 3.0.2
- Check compilation and linking status only when debug WebGL context is used to improve performance (#144)

### 3.0.1
- Add CORS setting to allow loading image from a different domain

## 3.0.0

Codebase/Build tooling improvements
- Replace wildcard exports with named exports in index.js
- ES6 Conformant code base: stage-2 extensions removed
- Webpack based build
- Multiple examples now work standalone
- Experimental tree-shaking support: dist and dist-es6 directories
- Dependency removal, including removal of `autobind-decorator` dependency
- Changed precommit hook from `husky` to `pre-commit`
- `shader-modules`, `shader-tools`, `shaders` shader module system added to `/experimental`
- `probe` moved to `/experimental`
- `webgl` folder now contains both webgl1 and webgl2 classes

Feature Improvements
- Performance query using EXT_disjoint_timer_query #121

Breaking Changes:
- BREAKING CHANGE: Move node IO (loadImage etc) out of main src tree
  and into `packages`. This allows luma.gl to drop a number of big dependencies.
  The node IO code may be published as a separate module later.

### 2.10.4
- FIX: Fix for glGetDebugInfo regression on Intel processors.

### 2.10.3
- FIX: Fix for glGetDebugInfo regression under Node in 2.10.2.
- FIX: Add "experimental.js" to exported "files" in package.json.

### 2.10.2
- FEATURE: Introduce experimental ShaderCache
- FIX: for glGetDebugInfo under Firefox (WEBGL_debug_renderer_info issue)
- CHANGE: Removes glslify as a dependency, apps that depend on glslify
  must add it to their own package.json.

### 2.10.1
- FIX: glslify path.

### 2.10.0
- Introduce new gl-matrix based math library.
- Move old math lib to deprecated folder.
- Move FBO to deprecated folder.
- Examples converted to ES6. AnimationLoop class updates.
- Add back persistence example
- WebGL type and constant cleanup
- Fix glTypeToArray and use clamped arrays by default

### 2.9.1 GLSL shader compiler error handling
  - FIX: GLSL shader compiler error parsing

### 2.9.0  TimerQuery, WebGL Extension doc, fix crash on Travis CI
  - Support EXT_disjoint_timer_query
  - Document luma.gl use of WebGL extensions.
  - Fix: context creation crash when WEBGL_debug_info extension was undefined
  - Add

### 2.8.0  Debug log improvements, import fix
  - Debug logs now print unused attributes more compactly, number formatting
    improved.
  - FIX: io import issue in 2.7.0

### 2.7.0 - Add ability to import luma without io
  - import "luma.gl/luma" will import luma without io functions
  - import "luma.gl/io" will import luma io functions only
  - omitting io functions significantly reduces dependencies

### 2.6.0 - "64 bit" camera projection matrix
 - Add 64 bit matrix to Luma.gl Camera
 - Updated linter rules

### 2.5.4 - FIX: Luma global initialization
- Makes the luma object available in console for debugging.
- Makes optional headless support more reliable.

### 2.5.3 - FIX: Linux rendering issues
- Add missing call to getAttribLocation.
- Some polish on luma's built-in attribute/uniform logging

### 2.5.2 - FIX: document.navigator override
- More gentle override, carefully restoring the variable.

### 2.5.1 - FIX: make deprecated AttributeManager.add updateMap work again
- Attribute manager changes

### 2.5.0 - Node.js/AttributeManager/Renderer/Program.render()/Examples

- Ensure luma.gl does not fail under node until createGLContext is called.

- Program.render() now takes a map of uniforms,
  reducing need to "set" uniforms before render.

- AttributeManager improvements
    - add logging/instrumentation hooks, to help apps profile attribute updates.
    - Pass AttributeManager.update() parameters through to the individual
      attribute updater funcs, enabling app to generate shared attributes
      independently of layers for additional performance gains.
    - Add JSDoc to all public methods and basic test cases.


src/webgl-utils/parameters.js
import GL from './constants';
import assert from 'assert';

// DEFAULT SETTINGS - FOR FAST CACHE INITIALIZATION AND CONTEXT RESETS

export const GL_PARAMETER_DEFAULTS = {
  [GL.BLEND]: false,
  [GL.BLEND_COLOR]: new Float32Array([0, 0, 0, 0]),
  [GL.BLEND_EQUATION_RGB]: GL.FUNC_ADD,
  [GL.BLEND_EQUATION_ALPHA]: GL.FUNC_ADD,
  [GL.BLEND_SRC_RGB]: GL.ONE,
  [GL.BLEND_DST_RGB]: GL.ZERO,
  [GL.BLEND_SRC_ALPHA]: GL.ONE,
  [GL.BLEND_DST_ALPHA]: GL.ZERO,
  [GL.COLOR_CLEAR_VALUE]: new Float32Array([0, 0, 0, 0]), // TBD
  [GL.COLOR_WRITEMASK]: [true, true, true, true],
  [GL.CULL_FACE]: false,
  [GL.CULL_FACE_MODE]: GL.BACK,
  [GL.DEPTH_TEST]: false,
  [GL.DEPTH_CLEAR_VALUE]: 1,
  [GL.DEPTH_FUNC]: GL.LESS,
  [GL.DEPTH_RANGE]: new Float32Array([0, 1]), // TBD
  [GL.DEPTH_WRITEMASK]: true,
  [GL.DITHER]: true,
  [GL.FRONT_FACE]: GL.CCW,
  [GL.GENERATE_MIPMAP_HINT]: GL.DONT_CARE,
  [GL.LINE_WIDTH]: 1,
  [GL.POLYGON_OFFSET_FILL]: false,
  [GL.POLYGON_OFFSET_FACTOR]: 0,
  [GL.POLYGON_OFFSET_UNITS]: 0,
  [GL.SAMPLE_COVERAGE_VALUE]: 1.0,
  [GL.SAMPLE_COVERAGE_INVERT]: false,
  [GL.SCISSOR_TEST]: false,
  // Note: Dynamic value. If scissor test enabled we expect users to set correct scissor box
  [GL.SCISSOR_BOX]: new Int32Array([0, 0, 1024, 1024]),
  [GL.STENCIL_TEST]: false,
  [GL.STENCIL_CLEAR_VALUE]: 0,
  [GL.STENCIL_WRITEMASK]: 0xFFFFFFFF,
  [GL.STENCIL_BACK_WRITEMASK]: 0xFFFFFFFF,
  [GL.STENCIL_FUNC]: GL.ALWAYS,
  [GL.STENCIL_REF]: 0,
  [GL.STENCIL_VALUE_MASK]: 0xFFFFFFFF,
  [GL.STENCIL_BACK_FUNC]: GL.ALWAYS,
  [GL.STENCIL_BACK_REF]: 0,
  [GL.STENCIL_BACK_VALUE_MASK]: 0xFFFFFFFF,
  [GL.STENCIL_FAIL]: GL.KEEP,
  [GL.STENCIL_PASS_DEPTH_FAIL]: GL.KEEP,
  [GL.STENCIL_PASS_DEPTH_PASS]: GL.KEEP,
  [GL.STENCIL_BACK_FAIL]: GL.KEEP,
  [GL.STENCIL_BACK_PASS_DEPTH_FAIL]: GL.KEEP,
  [GL.STENCIL_BACK_PASS_DEPTH_PASS]: GL.KEEP,
  // Dynamic value: We use [0, 0, 1024, 1024] as default, but usually this is updated in each frame.
  [GL.VIEWPORT]: new Int32Array([0, 0, 1024, 1024]),
  // WEBGL1 PIXEL PACK/UNPACK MODES
  [GL.PACK_ALIGNMENT]: 4,
  [GL.UNPACK_ALIGNMENT]: 4,
  [GL.UNPACK_FLIP_Y_WEBGL]: false,
  [GL.UNPACK_PREMULTIPLY_ALPHA_WEBGL]: false,
  [GL.UNPACK_COLORSPACE_CONVERSION_WEBGL]: GL.BROWSER_DEFAULT_WEBGL,

  // WEBGL2 / EXTENSIONS
  // gl1: 'OES_standard_derivatives'
  [GL.FRAGMENT_SHADER_DERIVATIVE_HINT]: GL.DONT_CARE,
  // RASTERIZER_DISCARD ...
  [GL.PACK_ROW_LENGTH]: 0,
  [GL.PACK_SKIP_PIXELS]: 0,
  [GL.PACK_SKIP_ROWS]: 0,
  [GL.UNPACK_ROW_LENGTH]: 0,
  [GL.UNPACK_IMAGE_HEIGHT]: 0,
  [GL.UNPACK_SKIP_PIXELS]: 0,
  [GL.UNPACK_SKIP_ROWS]: 0,
  [GL.UNPACK_SKIP_IMAGES]: 0
};

// SETTER TABLES - ENABLES SETTING ANY PARAMETER WITH A COMMON API

const enable = (gl, value, key) => value ? gl.enable(key) : gl.disable(key);
const hint = (gl, value, key) => gl.hint(key, value);
const pixelStorei = (gl, value, key) => gl.pixelStorei(key, value);

// NOTE: When value type is a string, it will be handled by 'GL_PARAMETER_COMPOSITE_SETTERS'

export const GL_PARAMETER_SETTERS = {
  [GL.BLEND]: enable,
  [GL.BLEND_COLOR]: (gl, value) => gl.blendColor(...value),
  [GL.BLEND_EQUATION_RGB]: 'blendEquation',
  [GL.BLEND_EQUATION_ALPHA]: 'blendEquation',
  [GL.BLEND_SRC_RGB]: 'blendFunc',
  [GL.BLEND_DST_RGB]: 'blendFunc',
  [GL.BLEND_SRC_ALPHA]: 'blendFunc',
  [GL.BLEND_DST_ALPHA]: 'blendFunc',
  [GL.COLOR_CLEAR_VALUE]: (gl, value) => gl.clearColor(...value),
  [GL.COLOR_WRITEMASK]: (gl, value) => gl.colorMask(...value),
  [GL.CULL_FACE]: enable,
  [GL.CULL_FACE_MODE]: (gl, value) => gl.cullFace(value),
  [GL.DEPTH_TEST]: enable,
  [GL.DEPTH_CLEAR_VALUE]: (gl, value) => gl.clearDepth(value),
  [GL.DEPTH_FUNC]: (gl, value) => gl.depthFunc(value),
  [GL.DEPTH_RANGE]: (gl, value) => gl.depthRange(...value),
  [GL.DEPTH_WRITEMASK]: (gl, value) => gl.depthMask(value),
  [GL.DITHER]: enable,
  [GL.FRAGMENT_SHADER_DERIVATIVE_HINT]: hint,
  [GL.FRONT_FACE]: (gl, value) => gl.frontFace(value),
  [GL.GENERATE_MIPMAP_HINT]: hint,
  [GL.LINE_WIDTH]: (gl, value) => gl.lineWidth(value),
  [GL.POLYGON_OFFSET_FILL]: enable,
  [GL.POLYGON_OFFSET_FACTOR]: 'polygonOffset',
  [GL.POLYGON_OFFSET_UNITS]: 'polygonOffset',
  [GL.SAMPLE_COVERAGE_VALUE]: 'sampleCoverage',
  [GL.SAMPLE_COVERAGE_INVERT]: 'sampleCoverage',
  [GL.SCISSOR_TEST]: enable,
  [GL.SCISSOR_BOX]: (gl, value) => gl.scissor(...value),
  [GL.STENCIL_TEST]: enable,
  [GL.STENCIL_CLEAR_VALUE]: (gl, value) => gl.clearStencil(value),
  [GL.STENCIL_WRITEMASK]: (gl, value) => gl.stencilMaskSeparate(GL.FRONT, value),
  [GL.STENCIL_BACK_WRITEMASK]: (gl, value) => gl.stencilMaskSeparate(GL.BACK, value),
  [GL.STENCIL_FUNC]: 'stencilFuncFront',
  [GL.STENCIL_REF]: 'stencilFuncFront',
  [GL.STENCIL_VALUE_MASK]: 'stencilFuncFront',
  [GL.STENCIL_BACK_FUNC]: 'stencilFuncBack',
  [GL.STENCIL_BACK_REF]: 'stencilFuncBack',
  [GL.STENCIL_BACK_VALUE_MASK]: 'stencilFuncBack',
  [GL.STENCIL_FAIL]: 'stencilOpFront',
  [GL.STENCIL_PASS_DEPTH_FAIL]: 'stencilOpFront',
  [GL.STENCIL_PASS_DEPTH_PASS]: 'stencilOpFront',
  [GL.STENCIL_BACK_FAIL]: 'stencilOpBack',
  [GL.STENCIL_BACK_PASS_DEPTH_FAIL]: 'stencilOpBack',
  [GL.STENCIL_BACK_PASS_DEPTH_PASS]: 'stencilOpBack',
  [GL.VIEWPORT]: (gl, value) => gl.viewport(...value),

  // WEBGL1 PIXEL PACK/UNPACK MODES
  [GL.PACK_ALIGNMENT]: pixelStorei,
  [GL.UNPACK_ALIGNMENT]: pixelStorei,
  [GL.UNPACK_FLIP_Y_WEBGL]: pixelStorei,
  [GL.UNPACK_PREMULTIPLY_ALPHA_WEBGL]: pixelStorei,
  [GL.UNPACK_COLORSPACE_CONVERSION_WEBGL]: pixelStorei,

  // WEBGL2 PIXEL PACK/UNPACK MODES
  // RASTERIZER_DISCARD ...
  [GL.PACK_ROW_LENGTH]: pixelStorei,
  [GL.PACK_SKIP_PIXELS]: pixelStorei,
  [GL.PACK_SKIP_ROWS]: pixelStorei,
  [GL.UNPACK_ROW_LENGTH]: pixelStorei,
  [GL.UNPACK_IMAGE_HEIGHT]: pixelStorei,
  [GL.UNPACK_SKIP_PIXELS]: pixelStorei,
  [GL.UNPACK_SKIP_ROWS]: pixelStorei,
  [GL.UNPACK_SKIP_IMAGES]: pixelStorei
};

const GL_PARAMETER_COMPOSITE_SETTERS = {
  blendEquation: (gl, values) => gl.blendEquationSeparate(
    values[GL.BLEND_EQUATION_RGB],
    values[GL.BLEND_EQUATION_ALPHA]
  ),
  blendFunc: (gl, values) => gl.blendFuncSeparate(
    values[GL.BLEND_SRC_RGB],
    values[GL.BLEND_DST_RGB],
    values[GL.BLEND_SRC_ALPHA],
    values[GL.BLEND_DST_ALPHA]
  ),
  polygonOffset: (gl, values) => gl.polygonOffset(
    values[GL.POLYGON_OFFSET_FACTOR],
    values[GL.POLYGON_OFFSET_UNITS]
  ),
  sampleCoverage: (gl, values) => gl.sampleCoverage(
    values[GL.SAMPLE_COVERAGE_VALUE],
    values[GL.SAMPLE_COVERAGE_INVERT]
  ),
  stencilFuncFront: (gl, values) => gl.stencilFuncSeparate(GL.FRONT,
    values[GL.STENCIL_FUNC],
    values[GL.STENCIL_REF],
    values[GL.STENCIL_VALUE_MASK]
  ),
  stencilFuncBack: (gl, values) => gl.stencilFuncSeparate(GL.BACK,
    values[GL.STENCIL_BACK_FUNC],
    values[GL.STENCIL_BACK_REF],
    values[GL.STENCIL_BACK_VALUE_MASK]
  ),
  stencilOpFront: (gl, values) => gl.stencilOpSeparate(GL.FRONT,
    values[GL.STENCIL_FAIL],
    values[GL.STENCIL_PASS_DEPTH_FAIL],
    values[GL.STENCIL_PASS_DEPTH_PASS]
  ),
  stencilOpBack: (gl, values) => gl.stencilOpSeparate(GL.BACK,
    values[GL.STENCIL_BACK_FAIL],
    values[GL.STENCIL_BACK_PASS_DEPTH_FAIL],
    values[GL.STENCIL_BACK_PASS_DEPTH_PASS]
  )
};

// GETTER TABLE - FOR READING OUT AN ENTIRE CONTEXT

const isEnabled = (gl, key) => gl.isEnabled(key);

// Exceptions for any keys that cannot be queried by gl.getParameters
export const GL_PARAMETER_GETTERS = {
  [GL.BLEND]: isEnabled,
  [GL.CULL_FACE]: isEnabled,
  [GL.DEPTH_TEST]: isEnabled,
  [GL.DITHER]: isEnabled,
  [GL.POLYGON_OFFSET_FILL]: isEnabled,
  [GL.SAMPLE_ALPHA_TO_COVERAGE]: isEnabled,
  [GL.SAMPLE_COVERAGE]: isEnabled,
  [GL.SCISSOR_TEST]: isEnabled,
  [GL.STENCIL_TEST]: isEnabled,

  // WebGL 2
  [GL.RASTERIZER_DISCARD]: isEnabled
};

// PUBLIC METHODS

// Sets any GL parameter regardless of function (gl.blendMode, ...)
// Needs `cache` parameter to fill in any missing values for composite setter functions
export function glSetParameters(gl, values, cache) {
  const compositeSetters = {};

  // Call primitive setters and make note of any composite setters
  for (const key in values) {
    const setter = GL_PARAMETER_SETTERS[key];
    // Composite setters should only be called once, so save them
    if (typeof setter === 'string') {
      compositeSetters[setter] = true;
    // only call setter if value has changed
    // TODO - deep equal on values?
    } else if (values[key] !== cache[key]) {
      // Note - the setter will automatically update this.state
      setter(gl, values[key], key);
    }
  }

  // Handle composite setters
  // Ensure that any non-provided values needed by composite setters are filled in from state
  const mergedValues = Object.assign({}, cache, values);
  for (const key in compositeSetters) {
    assert(cache);
    // TODO - avoid calling composite setters if values have not changed.
    const compositeSetter = GL_PARAMETER_COMPOSITE_SETTERS[key];
    // Note - the setter will automatically update this.state
    compositeSetter(gl, mergedValues);
  }
}

// Queries any GL parameter regardless of function (gl.getParameter/gl.isEnabled...)
export function glGetParameter(gl, key) {
  const getter = GL_PARAMETER_GETTERS[key];
  return getter ? getter(gl, key) : gl.getParameter(key);
}

// Copies the state from a context (gl.getParameter should not be overriden)
// Reads the entire WebGL state from a context
// Caveat: This generates a huge amount of synchronous driver roundtrips and should be
// considered a very slow operation, to be used only if/when a context already manipulated
// by external code needs to be synchronized for the first time
// @return {Object} - a newly created map, with values keyed by GL parameters
export function glCopyParameters(gl) {
  const state = {};
  for (const key in GL_PARAMETER_DEFAULTS) {
    state[key] = glGetParameter(gl, key);
  }
  return state;
}
